docs/investigations/input-validation/input-validation.md

```diff
@@ -0,0 +1,254 @@
+# Input Validation
+
+## Proposed validation libraries
+
+### go-playground/validator
+
+https://github.com/go-playground/validator - in this library developer uses struct tags to define validation rules. It is possible to define custom rules or struct validation rules. Validator instance caches struct info, so it has to be passed as dependency (a single instance of validator could be constructed in root resolver and passed to services. In this approach validation rules (tags) would be stored with model but the validators (validation logic) should be held outside the package, for example in `pkg`.
+
+### go-ozzo/ozzo-validation
+
+https://github.com/go-ozzo/ozzo-validation - this is an alternative library which isn't based on struct tags (so it's less error prune). This library allows for stateless validation and keeping custom validators logic close to the model.
+
+#### Example of application input validation:
+
+- Model
+```
+type ApplicationCreateInput struct {
+	Name           string
+	Description    *string
+	Labels         map[string]interface{}
+	HealthCheckURL *string
+	Webhooks       []*WebhookInput
+	Apis           []*APIDefinitionInput
+	EventAPIs      []*EventAPIDefinitionInput
+	Documents      []*DocumentInput
+}
+
+func (i *ApplicationCreateInput) Validate() error {
+	return validation.ValidateStruct(i,
+		validation.Field(&i.Name, validation.Required, validation.Length(0, 36), validation.By(inputvalidation.ValidateName)),
+		validation.Field(&i.HealthCheckURL, is.URL, validation.Length(0, 256)),
+		validation.Field(&i.Webhooks, validation.Each(validation.Required)),
+		validation.Field(&i.Apis, validation.Each(validation.Required)),
+		validation.Field(&i.EventAPIs, validation.Each(validation.Required)),
+		validation.Field(&i.Documents, validation.Each(validation.Required)),
+	)
+}
+
+type ApplicationUpdateInput struct {
+	Name           string
+	Description    *string
+	HealthCheckURL *string
+}
+
+func (i *ApplicationUpdateInput) Validate() error {
+	return validation.ValidateStruct(i,
+		validation.Field(&i.Name, validation.Required, validation.Length(0, 36), validation.By(inputvalidation.ValidateName)),
+		validation.Field(&i.HealthCheckURL, is.URL, validation.Length(0, 256)),
+	)
+}
+```
+
+- Custom validator function kept in pkg
+```
+func ValidateName(value interface{}) error {
+	s, ok := value.(string)
+	if !ok {
+		return errors.New("type has to be string")
+	}
+	if errorMsg := validation.NameIsDNSSubdomain(s, false); errorMsg != nil {
+		return errors.Errorf("%v", errorMsg)
+	}
+	return nil
+}
+
+```
+
+## Validation rules explanation
+
+- `name` - Kubernetes resources can have names up to 253 characters long. The characters allowed in names are: digits (`0`-`9`), lower case letters (`a`-`z`),`-`, and `.`.  We could use the same pattern to validate our names.
+- `required` - Cannot be nil or empty.
+- `not_empty` - Cannot be empty (can be nil if pointer).
+- `url` - Valid URL.
+- `max` - Maximal allowed length.
```

Comment:
I would go with `max_length` as we should use also `max` and `min` as validation rules for numeric values. Such numeric validation may be applied, for example, in values like port number validation which needs to be in between 0 or 1 and 65535.

```suggestion
- `max_length` - Maximal allowed length.
```

Reply:
I agree, but these aliases are only used here to make the table more readable, they are not used in code.